src/decode/lzma.rs

```diff
@@ -237,23 +260,38 @@ where
         &mut self,
         rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
     ) -> error::Result<()> {
-        loop {
-            if let Some(unpacked_size) = self.unpacked_size {
-                if self.output.len() as u64 >= unpacked_size {
-                    break;
-                }
-            } else if rangecoder.is_finished_ok()? {
-                break;
-            }
+        self.process_mode(rangecoder, Mode::Finish)
+    }
 
-            let pos_state = self.output.len() & ((1 << self.pb) - 1);
+    #[cfg(feature = "stream")]
+    pub fn process_stream<'a, R: io::BufRead>(
+        &mut self,
+        rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
+    ) -> error::Result<()> {
+        self.process_mode(rangecoder, Mode::Run)
+    }
 
-            // Literal
-            if !rangecoder.decode_bit(
-                // TODO: assumes pb = 2 ??
-                &mut self.is_match[(self.state << 4) + pos_state],
-            )? {
-                let byte: u8 = self.decode_literal(rangecoder)?;
+    /// Process the next iteration of the loop.
+    ///
+    /// If the update flag is true, the decoder's state will be updated.
+    ///
+    /// Returns true if we should continue processing the loop, false otherwise.
```

Comment:
I think an enumeration would be clearer than a boolean to express the result. Something like `enum { Continue, Finished }`?